Improving Functional Test With `pytest` Fixtures

axiomcura opened the pull request:
_modified from [EmbeddedArtistry](https://embeddedartistry.com/blog/2017/08/04/a-github-pull-request-template-for-your-projects/)_

# Description

Thank you for your contribution to pycytominer!
Please _succinctly_ summarize your proposed change.
What motivated you to make this change?

Please also link to any relevant issues that your code is associated with.

---
This  PR introduces using [`pytest`](https://docs.pytest.org/en/7.4.x/) fixtures

The main goal of this PR is to leverage fixtures to make functional test much easier. Below is a diagram on how fixtures and helper functions are implemented in a single test. 

![functional-testing-diagram drawio](https://github.com/WayScience/CytoSnake/assets/31600622/8a9346a2-ba73-46a6-8d70-7360afb8dfdc)
> Simple diagram illustrating how functional tests work: starting for the right side, (red) is the fixture used to generate temporary testing directories per test, (yellow) is the functional test with all the steps done, (green) is the a helper function known as `prepare_data()` that does all the heavy lifting to transporting datasets into the testing folder. 


The diagram illustrates the implementation of `CytoSnake`'s functional tests. Initially, a fixture (depicted in red) serves as input for the functional test (depicted in yellow). Within this process, `pytest` identifies that the fixture includes a path where the test should occur, hence generating a temporary directory. Additionally, the fixture's teardown function ensures that `pytest` removes the testing directory once the test finishes. Overall, leveraging these fixtures eliminates the need for developers to manually create and delete temporary directories for every test.

Additionally,  a new function called `prepare_dataset()` has been added. This function primarily serves to facilitate the easy selection and transfer of data to a testing directory. It requires just one input: the desired dataset's name. The dataset names correspond to the directory names located within `cytosnake/tests/functional/dataset`.

## Additional changes 

**New module**
- New module created `test_utils` for storing general helper functions, making testing development much easier. 

**Documentation changes**
- The new `test_utils` module has now been added into the RTD documentation website source code


## Future PR

A future pull request will cover the rationale, motivations, and the design process behind the implementation of functional tests in `CytoSnake`. This PR will also provide a standardized guide detailing how to effectively document functional tests on the Read the Docs (RTD) website.

---

## What is the nature of your change?functional

- [ ] Bug fix (fixes an issue).
- [x] Enhancement (adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [x] This change requires a documentation update.

# Checklist

Please ensure that all boxes are checked before indicating that a pull request is ready for review.

- [x] I have read the [CONTRIBUTING.md](CONTRIBUTING.md) guidelines.
- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [x] New and existing unit tests pass locally with my changes.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have deleted all non-relevant text in this pull request template.

MikeLippincott reviewed (commented):
Great PR. The documentation in the code looks amazing. The `.cli.md` files are above me so I can't provide much feedback... if you have time and could explain in person, that would be great! Thank you

axiomcura commented on cytosnake/tests/test_utils.py:
```diff
@@ -0,0 +1,31 @@
+
+    # returns exception name, refer to function documentation to understand
+    # the order of parsing the traceback to obtain exception name.
+    return traceback.splitlines()[-1].split(":")[0].split(".")[-1]
```
Fixed! https://github.com/WayScience/CytoSnake/pull/75/commits/fc0614cb642ab748b302ca8fada2629958c30da5

d33bs reviewed (commented):
Nice work! I left a few comments and suggestions throughout this review. Please don't hesitate to let me know if you have any questions or concerns.

I'm marking this as a "comment" review for now as I wasn't able to successfully run tests from the changes in this PR (running into a `FileNotFoundError` when running `pytest` for tests `test_multiplates_with_multi_platemaps`, `test_one_plate_one_platemap`, and `test_multiplate_maps_no_barcode`).

d33bs commented on cytosnake/tests/functional/test_cli.py:
```diff
@@ -1,226 +1,330 @@
+            meta_data_path[0] if len(meta_data_path) == 1 else meta_data_path
+        )
+
+        # get barcode
+        barcode_path = [
+            str(fpath.name) for fpath in all_files if fpath.suffix == ".txt"
+        ]
```
I don't have as much experience with barcode files yet but wonder if [this](https://github.com/WayScience/Benchmarking_NF1_data/blob/main/CellProfiler_pipelines/Metadata/barcode_platemap.csv) might be an example from another context which helps (showing the file potentially in `.csv` form). Maybe another way of considering this comment would be to think about what value the if-branch in the comprehension adds; if it were removed would you retain the functionality you seek and also simplify things? If it yields an error downstream on removal it could be a good indication of necessary development work.

axiomcura replied:
Just spoke with Greg. In order to prevent over engineering, we're going to start narrow (.csv and .txt fies). If there happens to be large diversity of different formats in barcodes, then we would have to think of another approach to validate barcode files in CytoSnake

axiomcura commented on cytosnake/tests/functional/datasets/standard_sqlite_single/metadata/platemap/platemap1.csv:
That's a good point. I think that's a great idea! When I was developing this, I had no idea how the datasets should be implemented because the unit and workflow tests were not developed yet.

axiomcura commented on cytosnake/tests/__init__.py:
After reading this. Tests are totally meant for development; therefore, the `test` directory will be in the root directory folder.

axiomcura commented on cytosnake/tests/functional/test_cli.py:
```diff
@@ -1,226 +1,330 @@
+# --------------
+# helper functions
+# --------------
+@dataclass
+class DataFiles:
```
I made this dataclass for tests only. I don't see it being used for file handling in CytoSnake at the moment.I'll move this into the `test_utils` module for now.

axiomcura replied:
moved to `test_utils`: 8145a91a60a00a35b7c8f281ce417eecec2e4f50

axiomcura commented on cytosnake/tests/functional/test_cli.py:
```diff
@@ -1,226 +1,330 @@
-contains.
+The test scenarios are categorized into two main types:
+
+Positive Cases:
```
added: 1cc1d23b1134804423acc8ea36410167ef625396

axiomcura commented on cytosnake/tests/functional/test_cli.py:
```diff
@@ -1,226 +1,330 @@
+
+    # execute
+    cmd = f"cytosnake init -d *plate -m metadata -b {barcode}".split()
+    proc = subprocess.run(cmd, capture_output=True, text=True, check=False)
```
added: 8145a91a60a00a35b7c8f281ce417eecec2e4f50

axiomcura commented on cytosnake/tests/functional/test_cli.py:
```diff
@@ -1,226 +1,330 @@
-        self.tmp_path = tmp_path
+# --------------
+# helper functions
+# --------------
```
added: 8145a91a60a00a35b7c8f281ce417eecec2e4f50

axiomcura commented on cytosnake/tests/functional/test_cli.py:
```diff
@@ -1,226 +1,330 @@
+        plate_data = [
+            str(fpath.name)
+            for fpath in all_files
+            if fpath.suffix == ".parquet" or fpath.suffix == ".sqlite"
```
added:  cc439b5e6134d62e40254dcf6d8e808675d5e1b6

axiomcura commented on docs/cytosnake.tests.md:
```diff
@@ -0,0 +1,21 @@
+# cytosnake.tests package
+
```
Yes, this documentation about the tests will be added in the next PR!

axiomcura commented on cytosnake/tests/functional/test_cli.py:
```diff
@@ -1,226 +1,330 @@
-    tmpdir_metadata_path = test_dir / metadata_dir_name / "platemap"
-    tmpdir_metadata_path.mkdir(exist_ok=True, parents=True)
+    # get testing dataset_path
+    data_dir_path = pathlib.Path("./datasets")
```
added:  ee8bd553cf6b925c30d7ff5a4abcd334084dcdf5

axiomcura commented on cytosnake/tests/functional/test_cli.py:
```diff
@@ -1,226 +1,330 @@
+    proc = subprocess.run(cmd, capture_output=True, text=True, check=False)
+
+    # assert checks
+    assert proc.returncode == 0
```
added: ee8bd553cf6b925c30d7ff5a4abcd334084dcdf5

axiomcura commented on cytosnake/tests/functional/datasets/standard_sqlite_single/plate_data1.sqlite:
added!

axiomcura commented on cytosnake/tests/functional/test_cli.py:
```diff
@@ -1,226 +1,322 @@
-    # leave test directory
-    os.chdir(test_module)
+    # Grab raised exception
+    raised_exception = get_raised_error(proc.stderr)
```
@d33bs Mind giving this a second round of review?

Here are some changes that I have done:
- I have moved the testing folder outside the module and placed it into the project root folder
- I removed the zero based data files and added NF1 dataset
- dataset folder is outside the functional test. 
- test_utils module has been moved into the `utils` section in CytoSnake
- added more assertions for my robust testing 
-

axiomcura commented:
@d33bs  Mind giving this a second round of review?

Here are some changes that I have done:
- I have moved the testing folder outside the module and placed it into the project root folder
- I removed the zero based data files and added NF1 dataset
- dataset folder is outside the functional test. 
- test_utils module has been moved into the `utils` section in CytoSnake
- added more assertions for my robust testing 
-

d33bs commented on tests/functional/test_cli.py:
```diff
@@ -0,0 +1,227 @@
+    test_utils.prepare_dataset(test_data_name="nf1-data", test_dir_path=testing_dir)
+
+    # Grab raised exception
+    with pytest.raises(subprocess.CalledProcessError) as subproc_error:
```
🙌

d33bs commented on tests/functional/test_cli.py:
```diff
@@ -0,0 +1,227 @@
+    """
+
+    # transfer data to testing folder
+    test_utils.prepare_dataset(test_data_name="nf1-data", test_dir_path=testing_dir)
```
Seeing how this pattern repeats, would this be a good opportunity for a fixture?

axiomcura replied:
I was thinking about this earlier and I feel like the design of using fixture that require parameters are a bit daunting and may be not easy for new developers. 

Below is an example if `prepare_dataset()` becomes a fixture:
```python
@pytest.mark.parametrize("prepare_data", ["nf1"], indirect=True)
def text_example(testing_dir, prepare_dataset) -> None:
    # testing code below
```

However, if a redesign of the testing framework is conducted, then I will consider this approach.

d33bs commented on tests/datasets/nf1-data-nobarcode/metadata/platemap/platemap_NF1_plate1.csv:
Did this and the related files come from other locations (and include efforts from other people)? If so, consider adding a reference of some kind to help document their contributions to this project (for example, through a [CITATION.cff](https://citation-file-format.github.io/) file using the `references` key).

axiomcura replied:
added: 1c6286ddd2a7f3dde62b1bce818a4a90acae1268

d33bs commented on tests/functional/pytest.ini:
```diff
@@ -0,0 +1,4 @@
+[pytest]
+markers =
+    postive: marks for postive test
```
```suggestion
    positive: marks for positive test
```

d33bs reviewed (approved):
Great work in addressing the comments and making changes! I've left a few more comments and generally felt things looked good. Please don't hesitate to let me know if you have any questions.

axiomcura commented on tests/functional/test_cli.py:
```diff
@@ -0,0 +1,227 @@
+    all_plates = list(data_folder.glob("*.sqlite"))
+
+    # assert checks
+    assert proc.returncode == 0
+    assert data_folder.exists()
+    assert cytosnake_file.exists()
+    assert metadata_in_datafolder.exists()
+    assert all([str(plate_data).endswith(".sqlite") for plate_data in all_plates])
```
Unfortunately, no, those tests will be done within the either unit testing or workflow testing. The functional tests just focus on emulating user experience with the CLI. For example, setting up your directory to be a project directory. No files are being open and read into memory in this step.

axiomcura commented on tests/functional/pytest.ini:
done: 7b1291f86f39eeb19503ab59cf4900aa41c37632

axiomcura commented on tests/functional/test_cli.py:
```diff
@@ -0,0 +1,227 @@
+    assert cytosnake_file.exists()
+    assert barcodes_in_datafolder.exists()
+    assert metadata_in_datafolder.exists()
+    assert all([str(plate_data).endswith(".sqlite") for plate_data in all_plates])
```
done! f2cc1761d30fba3a6cdcd919749500c50421df18

axiomcura commented:
@d33bs I have added comments and suggestions. I'll be merging until the end of the day if you do not have any more comments or suggestions. 

Thank you!
